### DISCUSSION TITLE
Packing solution as managed

### DISCUSSION THREAD

**Discussion by moizhraj** [microsoft/powerplatform-build-tools#710] [Q&A]
Hi, I have a extracted dump from a dev environment for version control in ADO. I am using Power Platform Build tools for Azure DevOps to pack this dump to a managed solution. Since the extracted dump is an unmanaged one, when i try to specify SolutionType as "Both" or even "Managed" in the PP Pack Solution task, it fails with an error as "Solution package type did not match requested type." A workaround is that I pack this solution as unmanaged which works, then import it to a staging env, and then import it back as  a managed solution to be deployed to PPE or Prod. Appreciate any help if there is a better way to do it.

**Comment by parvezghumra** [comment]
@moizhraj In order to enable solutions to be packed as managed you need to have the managed unpacked files available. If you don't have these then you will not be able to pack as managed/both. My suggestion would be to ensure that when you export and unpack your solution from your development environment, you do this for both a managed and unmanaged solution. This will enable you to repack both types as part of a build process

**Reply by moizhraj** [reply]
Thanks for the response. So I am not exporting and unpacking a solution from my PA development environment in my build pipeline. I have manually exported the unmanaged working solution as a base for my ADO repository. Now whatever changes I make, its in this repository only. And using this repo, I am trying to create a managed solution package. When we manually run the Solution Builder on this unmanaged zip, the command works fine but the same command run through the Power Platform Tools Task doesnt generate it. This is either a bug or intentional I am not sure but this is a much needed step..

**Reply by parvezghumra** [reply]
@moizhraj Could you share what the contents of your repository look like and how are you making changes to it's contents?

**Reply by sanartham** [reply]
I am implementing the similar strategy and i face the same issue.
High level steps:
1. After the development is done in Dev env, Developer is going to export the solution manually.
2. Manually Unpack it to the local Repo directory
3. Manually review the changes in VS -> Git Changes
4. Commit and Push
5. Kick-off the Azure DevOps Pipeline

Here is how my Repo looks like:

![image](https://github.com/microsoft/powerplatform-build-tools/assets/3372662/0a094628-5c0c-453c-b4df-b6467f8ab3a0)

**Reply by parvezghumra** [reply]
@moizhraj @sanartham As suggested above if you only export an unmanaged solution from your source environment, when you unpack it the resulting files can only be packed to an unmanaged solution. You must export and unpack both unmanaged and managed solutions in order to be able to pack both types of solutions from the resulting files later.

**Reply by devkeydet** [reply]
In order to build either unmanaged or managed from source control, you need to export and unpack both unmanaged and managed.   This is because there are still some files that have a different unpacked format in the unmanaged vs managed state.  There are a few ways to go about this.  You could do all the steps manually.

-Export Unmanaged
-Export Managed (same directory as unmanaged)
-Unpack both using the "Both" option

Here's an example:
`pac solution export --name BasicGallery --path solution-export-temp --managed true --overwrite true`
`pac solution export --name BasicGallery --path solution-export-temp --managed false --overwrite true`
`pac solution unpack --zipfile solution-export-temp/BasicGallery.zip --packagetype Both --folder unpacked-solution-both`

Once you've done that, you can `pac solution pack` with a `--packagetype` of your preference, including `Both`:
`pac solution pack --folder unpacked-solution-both --packagetype Both --zipfile packed-solution-both/BasicGallery.zip`

However, pac has a nice feature in `pac solution sync` and `pac solution clone` which will effectively do all of this for you.  `pac solution clone` is intended to initialize this process the first time.  `pac solution sync` is for subsequently bringing down the latest from your environment.  It also creates a *.cdsproj file, which you can run `msbuild` or `dotnet build` on and it will build both of your solutions.  The following command will automate all of it for the first time.
`pac solution clone --name BasicGallery --packagetype Both`  

If you then go into the *.cdsproj and update the `SolutionPackageType` in the XML by uncommenting the following:
![image](https://github.com/microsoft/powerplatform-build-tools/assets/4499296/6185a92f-86bf-49ab-acf2-73f922abbac4)
...and updating it to:
![image](https://github.com/microsoft/powerplatform-build-tools/assets/4499296/6f909202-07c8-4a6d-812c-f8b3e3b46c86)
...you can run `msbuild` or `dotnet build` in the folder where the *.cdsproj resides.  The result will be packed solutions file that are both managed and unmanaged.
![image](https://github.com/microsoft/powerplatform-build-tools/assets/4499296/97905f9a-c55b-4f1a-ac18-355978753f7a)
If you go this route, you also have the added benefit of being able to use `pac solution add-reference` to pcf and plugin dll projects.  By doing so *.cdsproj will build code first components first, then the Solutions that contain them.  I have an example of this more advanced scenario here.  
https://youtu.be/wzHLWNWcY2Q?si=WOQX5rLGF1j6Zik-
The difference is I do not show `pac solution clone` because I create the solution from scratch "code first" using `pac solution init`

**Reply by mannybellas** [reply]
> This is because there are still some files that have a different unpacked format in the unmanaged vs managed state.

hi @devkeydet Can you provide more details as to what files have different unpackaged format, what the differences are and why we need these to be different?  In many cases I notice that the files are exactly the same

**Reply by devkeydet** [reply]
I'll let the solutions team know about this discussion but will defer to them to comment.

**Reply by StylezDevops** [reply]
> I'll let the solutions team know about this discussion but will defer to them to comment.

any news?

**Comment by StylezDevops** [comment]
has there been any movement on this. This requirement - of course - is easy to get around by exporting and unpacking both to source control. however, am facing a situation where only unmanaged zips exist in a repo and now have to go through the whole process of importing them and then exporting them.. simply to get managed versions - rather than being able to set a flag with pac cli... which would make a wonderful feature.. being able to unpack and unmanaged zip and deploy it as either managed or unmanaged from pac cli - surely the pac cli team could work with the solutions team to understand what needs to change within the structure of the source to deploy as managed. 
@devkeydet 
fingers crossed.
